crates/sui-graphql-rpc/src/types/transaction_block_effects.rs

```diff
@@ -105,19 +115,87 @@ impl TransactionBlockEffects {
+            if let Some(Command::MoveCall(ptb_call)) = self
+                .programmable_transaction()?
+                .and_then(|ptb| ptb.commands.into_iter().nth(*command_idx))
+            {
+                let module_new = module.clone();
+                // Resolve the runtime module ID in the Move abort to the storage ID of the package
+                // that the abort occured in. This is important to make sure that we look at the
+                // correct version of the module when resolving the error.
+                *module = resolver
+                    .resolve_module_id(module_new, ptb_call.package.into())
+                    .await
+                    .map_err(|e| Error::Internal(format!("Error resolving Move location: {e}")))?;
+            }
+        }
+
+        match status {
+            NativeExecutionStatus::Success => Ok(None),
 
             NativeExecutionStatus::Failure {
                 error,
                 command: None,
-            } => Some(error.to_string()),
+            } => Ok(Some(error.to_string())),
 
             NativeExecutionStatus::Failure {
                 error,
                 command: Some(command),
             } => {
+                let error = 'error: {
+                    let ExecutionFailureStatus::MoveAbort(loc, code) = &error else {
+                        break 'error error.to_string();
+                    };
+                    let fname_string = if let Some(fname) = &loc.function_name {
+                        format!(" in function '{}' ", fname)
+                    } else {
+                        " ".to_string()
+                    };
+                    match resolver
+                        .resolve_clever_error(loc.module.clone(), *code)
+                        .await
+                    {
+                        Some(CleverError::CompleteError {
+                            module_id,
+                            error_constant,
+                            source_line_number,
+                            error_identifier,
+                        }) => {
+                            let const_str = error_constant.unwrap_or_else(FBase64::encode);
+                            format!(
+                                "'{error_identifier}' from module '{}'{fname_string}at source line {source_line_number}\n{const_str}",
+                                module_id.to_canonical_display(true)
+                            )
+                        }
+                        Some(CleverError::LineNumberOnly {
+                            module_id,
+                            source_line_number,
+                        }) => {
+                            format!(
+                                "Error from module '{}'{fname_string}at source line {source_line_number}",
+                                module_id.to_canonical_display(true)
+                            )
+                        }
+                        None => error.to_string(),
```

Comment:
Out of interest, what does the existing `ToString` give us for a non-clever error abort code? Does it degrade cleanly from what we have here?

Reply:
It does -- it will degrade to the standard display of the Move abort code within `ExecutionFailureStatus`. This is since `thiserror` which we use for that implements `Display`: https://docs.rs/thiserror/latest/thiserror/#details

Reply:
My question was more about whether the human-readable string we have written for that case looks like it's related to the one we get for the clever error case, (so if you imagine there's a continuum where on one extreme you have a clever error with all its fields provided, and on the other side you have a plain old legacy move abort, you can see from the error message that they all look like they belong to the same family of error).

Reply:
Ah I see! In that case I would say not so much, so I should/will update that.